### PR TITLE
pping: Use libxdp to load and attach XDP program

### PR DESCRIPTION
**Pull request by tohojo**
Since we now have libxdp as a submodule, we can switch pping over to using
libxdp for loading its XDP program. This requires switching the order of
attachment, because libxdp needs to be fed the BPF object in an unloaded
state (and will load it as part of attaching).

### CODE REVIEW

**Comment by tohojo** [comment]
@simosund PTAL :)

File: pping/pping.c [hunk 1]
```diff
@@ -76,6 +77,7 @@ struct pping_config {
 	char *event_map;
 	int xdp_flags;
```
Comment (simosund) [inline pping/pping.c:78]: As far as I can tell, xdp_flags is no longer used for anything, right? So it should probably be removed.

But a couple of questions related to what the XDP_FLAGS was used for (and could potentially be used for):

1. I'm not that familiar with libxdp, but if I recall correctly it let's you attach multiple XDP-programs to an interface (by chaining them with tail calls?). In that case, I guess we don't need any sort of replacement for no longer being able to "force" overwrite any existing XDP program on the interface. 

2. Right now I never let user's specify if "generic" xdp driver should be used or not (have just used the default behavior which falls back on generic if native is not supported). But if we think it's something that might be useful to add in the future, is there any way to send XDP_FLAGS to libxdp or some other way to affect which driver it tries to attach the XDP program with?

Comment (tohojo) [reply pping/pping.c:78]: > As far as I can tell, xdp_flags is no longer used for anything, right? So it should probably be removed.
> 
> But a couple of questions related to what the XDP_FLAGS was used for (and could potentially be used for):
> 
>     1. I'm not that familiar with libxdp, but if I recall correctly it let's you attach multiple XDP-programs to an interface (by chaining them with tail calls?). In that case, I guess we don't need any sort of replacement for no longer being able to "force" overwrite any existing XDP program on the interface.

Yup, exactly, 'force' is no longer necessary (and actively discouraged).

>     2. Right now I never let user's specify if "generic" xdp driver should be used or not (have just used the default behavior which falls back on generic if native is not supported). But if we think it's something that might be useful to add in the future, is there any way to send XDP_FLAGS to libxdp or some other way to affect which driver it tries to attach the XDP program with?

You can pass a mode to xdp_program__attach(); I currently pass `XDP_MODE_UNSPEC` which lets the kernel pick; but you can pass an explicit mode here instead.

I'll get rid of the flags field for now, thanks for the review :)

**Comment by tohojo** [comment]
There's no rush to merge this, just figured I'd do the conversion now that I finally got around to merging the libxdp support into the repository. So take your time and have a closer look if you want :)

**Comment by tohojo** [comment]
Yeah, never did get that error message resolved upstream. Guess I'll take another crack at it. Will merge this in the meantime so it doesn't get stale once you start moving things around :)

**Pull request merged**